[FLINK-4127] Check API compatbility for 1.1 in flink-core

rmetzger opened the pull request:
I checked all the newly introduced methods in public APIs by going through the reports generated from japicmp.
I've also put the reports (before my PR) into the JIRA: https://issues.apache.org/jira/browse/FLINK-4127

I added the new configuration parameters to the documentation, and renamed some new configuration keys.

@uce @tillrohrmann @mxm: What do you think about the renaming of the configuration keys?

greghogan commented on docs/setup/config.md:
```diff
@@ -85,6 +85,8 @@ The default fraction for managed memory can be adjusted using the `taskmanager.m
 
 - `taskmanager.memory.preallocate`: Can be either of `true` or `false`. Specifies whether task managers should allocate all managed memory when starting up. (DEFAULT: false)
 
+- `taskmanager.runtime.large-record-handler`: Whether to use the LargeRecordHandler when spilling. This feature is experimental. (DEFAULT: false)
```
Have we documented when it would be useful to enable `LargeRecordHandler`?

uce replied:
That's a very good point, Greg. I think we didn't. As it stands, the large record handler seems to have known issues. Personally, I would therefore discourage people from using it, but it totally makes sense to add a paragraph about what it actually does.

rmetzger replied:
@aljoscha, can you explain what the large record handler is doing?

aljoscha replied:
Not really, I'm afraid. I just recently disabled it by default because of the known issues.

What I can gather from the code is that it is a separate sort buffer that is intended for very large records. It is not obvious from the code what "very large records" are, however.

The known problem is that key serialization does not work correctly if the user specified a custom type or if Scala types are used because the `TypeAnalyzer` is used in the `LargeRecordHandler` to get a `TypeInformation` on the fly.

rmetzger replied:
Maybe we should not mention this in the documentation then?

aljoscha replied:
I would say so, yes.

mxm commented on flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java:
```diff
@@ -269,26 +269,26 @@
 	 * for other JVM memory usage.
 	 */
-	public static final String CONTAINERED_HEAP_CUTOFF_RATIO = "containered.heap-cutoff-ratio";
+	public static final String CONTAINER_HEAP_CUTOFF_RATIO = "container.heap-cutoff-ratio";
```
IMHO the `containered` prefix reflects the meaning of this config key better than `container`. It cuts off a fraction from the containerized Flink process, not from the container itself which the latter would imply.

rmetzger replied:
how about renaming the prefix to `containerized`

mxm replied:
I'm ok with that.

mxm commented on flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java:
```diff
@@ -269,26 +269,26 @@
 	 * Minimum amount of heap memory to remove in containers, as a safety margin.
 	 */
-	public static final String CONTAINERED_HEAP_CUTOFF_MIN = "containered.heap-cutoff-min";
+	public static final String CONTAINER_HEAP_CUTOFF_MIN = "container.heap-cutoff-min";
```
Same as above.

mxm commented on flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java:
```diff
@@ -269,26 +269,26 @@
 	 * in the flink-conf.yaml.
 	 */
-	public static final String CONTAINERED_MASTER_ENV_PREFIX = "containered.application-master.env.";
+	public static final String CONTAINER_MASTER_ENV_PREFIX = "container.application-master.env.";
```
How about we rename the config string to `container.master.env`?

rmetzger replied:
That makes sense. Application master is a YARN specific term afaik.

mxm commented on flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java:
```diff
@@ -269,26 +269,26 @@
 	 * setting custom environment variables for the workers (TaskManagers)
 	 */
-	public static final String CONTAINERED_TASK_MANAGER_ENV_PREFIX = "containered.taskmanager.env.";
+	public static final String CONTAINER_TASK_MANAGER_ENV_PREFIX = "container.taskmanager.env.";
```
How about we rename the config string to `container.worker.env`?

rmetzger replied:
That would be consistent with above renaming

rmetzger replied:
Actually, I'm not so sure about this anymore, because it taskManager is a name used in the documentation everywhere. Maybe we keep this name here?

tillrohrmann replied:
With this argument, we should also rename `CONTAINER_MASTER_ENV_PREFIX` to `container.jobmanager.env`, shouldn't we?

mxm commented:
Thanks for updating the documentation! I've made some suggestions regarding the names of the new configuration keys.

tillrohrmann commented on docs/setup/config.md:
```diff
@@ -107,12 +109,20 @@ Please make sure to set the maximum ticket life span high long running jobs. The
 
+### Resource Manager
+
+- `resourcemanager.rpc.port`: The config parameter defining the network port to connect to for communication with the resource manager.
```
Is it possible to specify a port range here?

mxm replied:
No, it doesn't. It is currently not in use (except for testing code). In Yarn, the port of the application master is always the port of the job manager because the two run in the same actor system.

tillrohrmann commented on docs/setup/config.md:
```diff
@@ -107,12 +109,20 @@ Please make sure to set the maximum ticket life span high long running jobs. The
+
 - `jobmanager.web.port`: Port of the JobManager's web interface (DEFAULT: 8081).
 
+- `jobmanager.web.tmpdir`: This configuration parameter allows defining the Flink web directory to be used by the web interface.
```
what is a web directory?

tillrohrmann commented on docs/setup/config.md:
```diff
@@ -230,8 +240,8 @@ definition. This scheme is used **ONLY** if no other scheme is specified (explic
 
-- `yarn.heap-cutoff-ratio`: (Default 0.25) Percentage of heap space to remove from containers started by YARN. When a user requests a certain amount of memory for each TaskManager container (for example 4 GB), we can not pass this amount as the maximum heap space for the JVM (`-Xmx` argument) because the JVM is also allocating memory outside the heap. YARN is very strict with killing containers w ...
-- `yarn.heap-cutoff-min`: (Default 384 MB) Minimum amount of memory to cut off the requested heap size.
+- `container.heap-cutoff-ratio`: (Default 0.25) Percentage of heap space to remove from containers started by YARN. When a user requests a certain amount of memory for each TaskManager container (for example 4 GB), we can not pass this amount as the maximum heap space for the JVM (`-Xmx` argument) because the JVM is also allocating memory outside the heap. YARN is very strict with killing contain ...
```
Maybe insert *, for example,* after YARN because the parameter is not specific to YARN, is it?

mxm replied:
+1

tillrohrmann commented:
Good changes @rmetzger.

rmetzger commented:
I addressed all comments. It seems to me that the configuration parameters for YARN are overly complicated now because they are separated in resource manager and YARN now.

mxm commented on flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java:
```diff
@@ -269,26 +269,26 @@
 	 * in the flink-conf.yaml.
 	 */
-	public static final String CONTAINERED_MASTER_ENV_PREFIX = "containered.application-master.env.";
+	public static final String CONTAINERIZED_MASTER_ENV_PREFIX = "containerized.master.env.";
```
I thought this should be `CONTAINER_MASTER_ENV_PREFIX = container.master.env.`?

rmetzger replied:
I renamed it.
I think we misunderstood each other. I thought our discussion regarding `container` / `containered` / `containerized` was about the prefix in general, not for the heap cutoff settings alone.

rmetzger commented:
The YARN settings are now spread across three prefixes:
- `yarn.`
- `containerized.`
- `container.`

I don't think this is helpful for making the system easy to configure.
What's your opinion on this @tillrohrmann @uce @StephanEwen ?

StephanEwen commented:
That would be good to address. Can we get away with only `yarn.` and `taskmanager.`?

StephanEwen commented:
There are a lot of config options that seem to exist in two variants: One for standalone setup, one for containered setup.

Why are we making this distinction?

tillrohrmann commented:
I think it's a good idea to merge `container` and `containerized`. 

I think the idea is to introduce some general container parameters, such as the `heap-cutoff-ratio`, which are valid not only for yarn but also for mesos. Therefore, the `container/containered/containerized` prefix was introduced. Given that we don't have mesos support yet it might be confusing. However, changing the parameter names too often, given that they are declared public evolving, will cause pain for users.

Which duplicate configuration parameters are you referring to @StephanEwen?

rmetzger commented:
I'll rename `containerized` to `container` once I got a third confirmation from a committer on this issue.

aljoscha commented:
Sorry for chiming in late, but can't - `heap-cutoff-ratio` and `heap-cutoff-min` be general parameters that apply to all VMs we start?

Same for `master.env.*` and `taskmanager.env.*`. Shouldn't this also apply to all VMs that we start?

rmetzger commented:
So you are suggesting to use the heap cutoff also for the JVM started in standalone mode?
The cutoff was introduced for YARN because there is a difference in the amount of used memory for a container we specify to YARN vs the memory the JVM is actually using.
So if we specify a container of 1000 MB, it will be killed by YARN as soon as we use 1001 MBs. For operating systems, this doesn't apply and if users run into issues, the can just set a lower heap size.
For Mesos, I don't know how it behaves, but I hope that there is a similar requirement, otherwise, we would have prematurely optimized the configuration keys.
I'm against applying this mechanism to all JVMs we start.

For the environment variables of the standalone mode, its quite easy to put the environment variables into the config.sh file, or just set the env variables.
But true, we could do a follow up JIRA to extract those env variables and set them in the config.sh script.

aljoscha commented:
My thinking was that Mesos maybe didn't require this, yes. Having it for Yarn only makes for a better user experience right now. So we could leave it under the "container" namespace.

rmetzger commented:
But with this argument, it would be even better to leave the configuration parameters as they are in 1.0 until we are 100% certain that mesos needs a similar mechanism.

aljoscha commented:
Yeah, then let's leave them as they were since we don't know yet what mesos will require there.

rmetzger commented:
All the container* config keys are now renamed to `containerized.`, so we have `yarn.` and `containerized.` as prefixes. But these config keys will be undocumented for now.
In the documentation, we still only list the `yarn.` configuration keys (and they are also picked up by the system).

I guess with the 1.2 release (when we are adding Mesos support) we can reconsider this decision and see how we are naming / grouping the configuration keys.

I'm going to merge this pull request after 24 hours.
